### ISSUE TITLE
generatingBalanceDepth is too small

### ISSUE THREAD

**Issue by bcdev-** [wavesplatform/Waves#42]
Default generatingBalanceDepth [in file NxtLikeConsensusModule.scala] is only 50.
Default in NXT is 1440. 50 is too small and could allow an attacker [with a miniscule stake] to hijack the network.

If this is just a temporary value for testnet, then it should be:
`generatingBalanceDepth = 1440 if mainnet else 50`

**Comment by bcdev-** [comment]
How could that compromise the network:
NXT consensus allows anyone to look "into the future" at who is generating blocks.
If an attacker guesses correctly who is going to generate every block for the next generatingBalanceDepth disaster happens.

Attack scenario:
An attacker predicts who is going to generate blocks for the next "generatingBalanceDepth" blocks. In other words he's predicting the value of a "calcGeneratorSignature" 49 blocks from now.
Using his predicted "calcGeneratorSignature", he calculates a Waves address that will generate block 50 with a microscopic stake [100 Waves would be enough IMO].
If his prediction was correct, he does the same thing for the next block... And the next...

If an attacker is successful in creating a chain of "generatingBalanceDepth" of addresses that are forging, the network now belongs to him - he's in a state of a permanent 51% attack, despite minuscule stake.


Predicting 50 blocks forward is not a hard task. 1440 on the other hand is nearly impossible.

**Comment by bcdev-** [comment]
I propose the following formula for consensus calculation:
```  private def calcGeneratorSignature(lastBlockData: NxtLikeConsensusBlockData, generator: PublicKeyAccount) =
    hash(lastBlockData.generationSignature ++ generator.publicKey ++ 1430BlocksAgoBlockData.hash)```

This way the attacker has to:
1. Guess all the forgers for 1440 blocks.
2. Generate the next 10 blocks by himself to know their hashes.

**Comment by asolovyov** [comment]
>> In other words he's predicting the value of "calcGeneratorSignature" 49 blocks from now.

In what way? If blockdata.generationSignature is not defined yet?

**Comment by bcdev-** [comment]
1. generationSignature is deterministic - it depends only on previous generationSignature&current account number
2. the list of active miners is well known from previous blocks
3. nxt forging is transparent - you know who will generate next block

All three of these combined can let you predict next blocks with high probability. 50 should be quite easy. 1440 not so much.

**Comment by asolovyov** [comment]
generationSignature of future blocks depends on future txs (to be included to these future blocks) and which are unknown at the moment of guess

**Comment by bcdev-** [comment]
No, it doesn't. That's a common misunderstanding, please consult the code if you're not convinced.

In fact _it would be a vulnerability_ if generationSignature depended on transactions. [A miner could calculate a block that allows him to mine the next block]

**Comment by bcdev-** [comment]
I want to make it depend on transactions, but they will be transactions that were executed 24 hours ago to make grinding impossible.

**Comment by asolovyov** [comment]
> No, it doesn't. 
Yes, I've checked, and you are right... (sign) 

Damn, I presumed the signature was f(txs, etc...) and hence nondeterministic... and TBH I'd prefer the grinding attack. If you are able to calculate the chain (no matter how long it is, 50 or 1440) and do it quickly (the FastCryptographicHash.hash is fast), and then _quickly_ calculate the next "address with a small stake", you will achieve your goal anyway - because you have to generate the "address with a small stake" only once, since it is a function of the +(1440 - 1) block.

**Comment by asolovyov** [comment]
The question here is complexity of the address calculation. This is brute force anyway, the question is what is preferable to have (harder to calculate), the grinding attack or this one?

And if so I don't believe that 1440 is a solution (see the previous comment). Please correct me if it is not so.

**Comment by asolovyov** [comment]
Ah, one note - the attack is possible if all balances remain unchanged, and for 1440 the probability is less than for 50. Then yes, the change makes sense.

**Comment by asolovyov** [comment]
"I want to make it depend on transactions, but they will be transactions that were executed 24 hours ago" - then generationSignature would be deterministic, as for the moment. But this is another story anyway, an orthogonal solution.

**Comment by bcdev-** [comment]
```
  private def calcGeneratorSignature(lastBlockData: NxtLikeConsensusBlockData, generator: PublicKeyAccount) =
    hash(lastBlockData.generationSignature ++ generator.publicKey ++ 1430BlocksAgoBlockData.hash)
```
With this formula you'd have to both:
1. Generate the next 10 blocks by yoursellf
2. Pray all 1430 miners align

Yes, it would be deterministic, and that's ok, because you cannot exploit it anyway. :-)

**Comment by asolovyov** [comment]
@bcdev- ok I see what you mean.

Looks like the PR has been applied, but let's clarify a couple of remaining items:
1. What is goal of the initially described attack? To generate one (the next) block?
2. If yes what is difference with the grinding attack? In both cases you look for / calculate a value matching a criteria, and the question is about complexity of such a calculation. 

As for 
```
    hash(lastBlockData.generationSignature ++ generator.publicKey ++ 1430BlocksAgoBlockData.hash)
```
solution, the grinding attack (to generate next block) is still possible, is not it? The question here is the calculation complexity. Again, the attack is possible unless it interferes with future transactions if they change balance of generating accounts.

**Comment by bcdev-** [comment]
The attack I'm describing is _not_ a grindingAttack. It's a prediction attack - you predict the generationSignature `balanceDepth` blocks in the future, you calculate an account with a low hit and you send 100 Waves there.
You do this every block.
If you're successful in predicting all `balanceDepth` blocks in a row, congratulations - you can now mine forever.

1. If this attack is successful, it will allow an attacker to **completely overtake the network** - meaning, they can generate infinite number of blocks by themselves. Essentially a 51% attack forever with a minuscule stake.
2. Griding attack requires computations and volatile blocks. If you can in any way influence your current block's `generationSignature`, that's called a grinding attack.

NXT is not susceptible to grinding attack, because you have no way of influencing `generationSignature`.
My formula is not susceptible to grinding attack because you have no way of influencing `generationSignature`. [It's taken from blocks 1440 and 1430 ago, not the current one]

**Comment by asolovyov** [comment]
@bcdev- 

> you predict the generationSignature balanceDepth blocks in the future

okay, this is very easy for the current consensus algo implementation, it is O(balanceDepth) i.e. linear

> [then] you calculate an account with a low hit 

What is complexity of this calculation? Is there a way to do that without plain brute-force (exhaustive) search? This is the main point...

**Comment by bcdev-** [comment]
Yes, it is a plain brute-force search, but it doesn't matter.
My CPU can do 70k signatures per second. That means that in 1 minute on average, I could generate an account that will have [this is a top-of-the-hat calculation, it's not the exact probability formula] `(70k*60)/2=2100000` advantage over an existing miner [on average will require 2100000* less waves than a typical forger]. So the computational requirements are extremely small.
Add to that a cloud of a 100 VPS'es, and you can calculate [in 1 minute] an account that will forge with 1 Waves if it gets elected 1440 blocks in the future.

The costs to do this attack are minuscule, both in terms of stake and power.

Adding `1430BlocksAgoBlockData.hash` will require an attacker to be a forger _who will forge the next 10 blocks_, which requires him to have a huge stake already. :-)

**Comment by asolovyov** [comment]
@bcdev- this makes sense... BTW you can try the approach on the mainnet in few days :-) I do not see any possible obstacle except volatility which should not be too big to prevent...

Okay, I give up! I don't see space for obvious attacks for your formula, at least I don't see now.

**Comment by bcdev-** [comment]
Writing an exploit for this attack is a waste of time IMO. It's better to just patch it in 5 minutes instead of wasting 40 hours proving it.

**Comment by asolovyov** [comment]
Taking into account the fact that 1430BlocksAgoBlockData depends on txs in the block, I am thinking about complex attack like:
1. Create a block to make hash(lastBlockData.generationSignature ++ generator.publicKey ++ 1430BlocksAgoBlockData.hash) be very small (the grinding attack step). Here 1430BlocksAgoBlockData is the last / new one.
2. 1430 blocks later perform the "predict" attack.

Indeed, at least two accounts are needed to perform the attack, and one of them should have a sufficient stake to perform step (1) 

But my brain hurts to think about all the details :-)

**Comment by asolovyov** [comment]
Hey @bcdev- 

The attack I was thinking about was overcomplicated and would not give much benefits (just a single block with a probability less than 100%).

So introduction of randomness you have suggested makes perfect sense despite we already have one (volatility of balances within 1000 blocks).

So yes, I agree that the formula for consensus calculation
```
  private def calcGeneratorSignature(lastBlockData: NxtLikeConsensusBlockData, generator: PublicKeyAccount) =
    hash(lastBlockData.generationSignature ++ generator.publicKey ++ (N Blocks Ago BlockData).hash)
```
where (Height - balanceDepth) < N < Height, e.g. N = 900 for current Waves impl 

looks better / more resistant than current one.

@sasha35625 @Tolsi @B83YPoj please consider.

@bcdev- is there a pull request for this?

**Comment by asolovyov** [comment]
@bcdev- it is being discussed now.

**Comment by bcdev-** [comment]
My opinion:
1. We raise the value to at least 1440, maybe even more. Better safe than sorry.
2. After 6 months implement my formula with a lower value. We could go as low as 500 with a considerable security margin.

Alternatively as "just to be sure" alternative we could simply monitor for [1] and if there is some suspect low-hit account 1440 blocks in the future, just don't mine [as waves] one block to throw it off. :-)
But raising the value is simpler.

**Comment by bcdev-** [comment]
Or even simpler solution for now: Waves project mines on several smaller accounts [instead of one big], *but* these accounts have 10% chance of skipping a block.
This would make it impossible to predict the skipped blocks, thus negating any potential attack.

Or we could just hope that nobody tries to do that. It's not a low-hanging-fruit type of attack, the probability is quite small. :-)
